# Tarea - Programación lineal

### Diseño de la Dieta Óptima

Se quiere producir comida para gatos de la manera más barata, no obstante se debe también asegurar que se cumplan los datos requeridos de analisis nutricional. Por lo que se quiere variar la cantidad de cada ingrediente para cumplir con los estandares nutricionales. Los requisitos que se tienen es que en $100$ gramos, se deben tener **por lo menos** $8$ gramos de proteína y $6$ gramos de grasa. Así mismo, no se debe tener más de $2$ gramos de fibra y $0.4$ gramos de sal.  

Los datos nutricionales se pueden obtener de la siguiente tabla:

Ingrediente|Proteína|Grasa|Fibra|Sal
:----|----|----|----|----
Pollo|  10.0%|08.0%|00.1%|00.2%
Carne|  20.0%|10.0%|00.5%|00.5%
Cordero|15.0%|11.0%|00.5%|00.7%
Arroz|  00.0%|01.0%|10.0%|00.2%
Trigo|  04.0%|01.0%|15.0%|00.8%
Gel|    00.0%|00.0%|00.0%|00.0%

Los costos de cada producto son:

Ingrediente|Costo por gramo
:----|----
Pollo|$\$$0.013
Carne|$\$$0.008
Cordero|$\$$0.010
Arroz|$\$$0.002
Trigo|$\$$0.005
Gel|$\$$0.001    

Lo que se busca optimizar en este caso es la cantidad de productos que se debe utilizar en la comida de gato, minimizando el costo total. Para simplificar la notación use las siguientes variables: 

+ $x_1:$ Gramos de pollo  
+ $x_2:$ Gramos de carne  
+ $x_3:$ Gramos de cordero  
+ $x_4:$ Gramos de arroz  
+ $x_5:$ Gramos de trigo  
+ $x_6:$ Gramos de gel

La tarea consiste en plantear el problemade programación lineal que permita satisfacer las necesidades alimenticias del gato al tiempo que minimice el costo total y resolverlo con `linprog`.

RESTRICCIONES:
- $x_{PP} + x_{PC} + X_{PCO} + X_{PA} + X_{PT} + X_{PG}\geq 8$
- $x_{GP} + x_{GC} + X_{GCO} + X_{GA} + X_{GT} + X_{GG}\geq 6$
- $x_{FP} + x_{FC} + X_{FCO} + X_{FA} + X_{FT} + X_{FG}\leq 2$
- $x_{SP} + x_{SC} + X_{SCO} + X_{SA} + X_{ST} + X_{SG}\leq 0.4$


$$\min 0.013x_{P} + 0.08x_{C} + 0.010x_{CO} + 0.002x_{A} + 0.005x_{T} + 0.001x_{G}$$


In [10]:
import numpy as np
import scipy.optimize as opt

c = np.array([0.013, 
              0.08, 
              0.010, 
              0.002, 
              0.005, 
              0.001])

Aub = np.array([ [-0.10, -0.20, -0.15, 0, -0.04, 0],
                 [-0.08, -0.10, -0.11, -0.01, -0.01, 0],
                 [0.001, 0.005, 0.005, 0.10, 0.15, 0],
                 [0.002, 0.005, 0.007, 0.002, 0.008, 0],
               ])

bub = np.array([-8,
                -6,
                 2,
                 0.40])

Aeq = np.array([[1,
                1,
                1,
                1,
                1,
                1]])

beq=np.array([
            [100]
])

opt.linprog (c, Aub, bub, Aeq, beq)

           con: array([0.])
 crossover_nit: 0
         eqlin:  marginals: array([0.001])
  residual: array([0.])
           fun: 0.5909090909090909
       ineqlin:  marginals: array([-0.        , -0.08181818, -0.        , -0.        ])
  residual: array([0.18181818, 0.        , 1.72727273, 0.01818182])
         lower:  marginals: array([0.00545455, 0.07081818, 0.        , 0.00018182, 0.00318182,
       0.        ])
  residual: array([ 0.        ,  0.        , 54.54545455,  0.        ,  0.        ,
       45.45454545])
       message: 'Optimization terminated successfully. (HiGHS Status 7: Optimal)'
           nit: 2
         slack: array([0.18181818, 0.        , 1.72727273, 0.01818182])
        status: 0
       success: True
         upper:  marginals: array([0., 0., 0., 0., 0., 0.])
  residual: array([inf, inf, inf, inf, inf, inf])
             x: array([ 0.        ,  0.        , 54.54545455,  0.        ,  0.        ,
       45.45454545])

El costo total sería 0.00545 y los gramos serían de la siguientes: 0.        , 0.        , 54.545455 gramos, 0.        , 0.        ,45.45 gramos        

### Problema de distribución de energía eléctrica

La Comisión Federal de Electricidad **(CFE)** dispone de cuatro plantas de generación para satisfacer la demanda diaria eléctrica en cuatro ciudades, Guadalajara, León y Morelia. Las plantas $1$, $2$ y $3$ pueden satisfacer $80$, $40$ y $60$ millones de KW al día respectivamente. Las necesidades de las ciudades de Guadalajara, León y Morelia son de $70$, $40$ y $70$ millones de Kw al día respectivamente. 


Los costos asociados al envío de suministro energético por cada millón de Kw entre cada planta y cada ciudad son los registrados en la siguiente tabla. 

-|Guadalajara|León|Morelia
:----|----|----|----
Planta 1|5|2|7
Planta 2|3|6|6
Planta 3|6|1|2

Y por último, las restricciones del problema, van a estar dadas por las capacidades de oferta y demanda de cada planta (en millones de KW) y cada ciudad.

Para simplificar la notación use las siguientes variables:

+ $x_1$: Kw (en millones) distribuidos de la Planta 1 a Guadalajara
+ $x_2$: Kw (en millones) distribuidos de la Planta 1 a León
+ $x_3$: Kw (en millones) distribuidos de la Planta 1 a Morelia
+ $x_4$: Kw (en millones) distribuidos de la Planta 2 a Guadalajara
+ $x_5$: Kw (en millones) distribuidos de la Planta 2 a León
+ $x_6$: Kw (en millones) distribuidos de la Planta 2 a Morelia
+ $x_7$: Kw (en millones) distribuidos de la Planta 3 a Guadalajara
+ $x_8$: Kw (en millones) distribuidos de la Planta 3 a León
+ $x_9$: Kw (en millones) distribuidos de la Planta 3 a Morelia

La tarea consiste en plantear el problema de programación lineal que permita satisfacer las necesidades de todas las ciudades al tiempo que minimice los costos asociados a la distribución y resolverlo con `linprog`.

Restricciones de producción:

- $x_{1G} + x_{1L} + x_{1M} \leq 80$
- $x_{2G} + x_{2L} + x_{2M} \leq 40$
- $x_{3G} + x_{3L} + x_{3M} \leq 60$

Restricciones de consumo:

- $x_{1G} + x_{2G} + X_{3G} = 70$
- $x_{1L} + x_{2L} + X_{3L} = 40$
- $x_{1M} + x_{2M} + X_{3M} = 70$

La función mínima:

$$\min 5x_{1G} + 2x_{1L} + 7x_{1M} + 3x_{2G} + 6x_{2L} + 6x_{2M} + 6x_{3G} + 1x_{3L} + 2x_{3M}$$


In [1]:
import numpy as np
import scipy.optimize as opt

c = np.array([5, 
              2, 
              7, 
              3, 
              6, 
              6,
              6,
              1,
              2])

Aub = np.array([
    [1, 1, 1, 0, 0, 0, 0, 0, 0 ],
    [0, 0, 0, 1, 1, 1, 0, 0, 0 ],
    [0, 0, 0, 0, 0, 0, 1, 1, 1 ],
               ])

bub = np.array([
                [80,
                 40,
                 60]
               ])
Aeq = np.array ([
    [1, 0, 0, 1, 0, 0, 1, 0, 0 ],
    [0, 1, 0, 0, 1, 0, 0, 1, 0],
    [0, 0, 1, 0, 0, 1, 0, 0, 1]
                ])

beq = np.array ([
                [70,
                40,
                70]
                ])

opt.linprog(c, Aub, bub, Aeq, beq)

           con: array([0., 0., 0.])
 crossover_nit: 0
         eqlin:  marginals: array([5., 2., 7.])
  residual: array([0., 0., 0.])
           fun: 540.0
       ineqlin:  marginals: array([-0., -2., -5.])
  residual: array([0., 0., 0.])
         lower:  marginals: array([0., 0., 0., 0., 6., 1., 6., 4., 0.])
  residual: array([30., 40., 10., 40.,  0.,  0.,  0.,  0., 60.])
       message: 'Optimization terminated successfully. (HiGHS Status 7: Optimal)'
           nit: 6
         slack: array([0., 0., 0.])
        status: 0
       success: True
         upper:  marginals: array([0., 0., 0., 0., 0., 0., 0., 0., 0.])
  residual: array([inf, inf, inf, inf, inf, inf, inf, inf, inf])
             x: array([30., 40., 10., 40.,  0.,  0.,  0.,  0., 60.])

El gasto sería de 540 millones y la producción sería de la siguiente manera. De la planta 1 en Guadalajara, se van a producir 30 millones de KW, en León 40 millones y en Morelia 10 millones. De la planta 2 en Guadalajara, se van a producir 40 millones de KW, en León y en Morelia ninguna unidad. De la planta 3 en Guadalajara, no se van a producir ningun millón, en León tampoco y en Morelia 60 millones. 